op-node/rollup/derive/engine_update.go

```diff
@@ -128,45 +195,77 @@ func confirmPayload(
 	updateSafe bool,
 	agossip async.AsyncGossiper,
 	sequencerConductor conductor.SequencerConductor,
+	builderClient builder.PayloadBuilder,
+	l2head eth.L2BlockRef,
+	metrics Metrics,
 ) (out *eth.ExecutionPayloadEnvelope, errTyp BlockInsertionErrType, err error) {
-	var envelope *eth.ExecutionPayloadEnvelope
+	var engineEnvelope *eth.ExecutionPayloadEnvelope
+	var builderEnvelope *eth.ExecutionPayloadEnvelope
 	// if the payload is available from the async gossiper, it means it was not yet imported, so we reuse it
 	if cached := agossip.Get(); cached != nil {
-		envelope = cached
+		engineEnvelope = cached
 		// log a limited amount of information about the reused payload, more detailed logging happens later down
 		log.Debug("found uninserted payload from async gossiper, reusing it and bypassing engine",
-			"hash", envelope.ExecutionPayload.BlockHash,
-			"number", uint64(envelope.ExecutionPayload.BlockNumber),
-			"parent", envelope.ExecutionPayload.ParentHash,
-			"txs", len(envelope.ExecutionPayload.Transactions))
+			"hash", engineEnvelope.ExecutionPayload.BlockHash,
+			"number", uint64(engineEnvelope.ExecutionPayload.BlockNumber),
+			"parent", engineEnvelope.ExecutionPayload.ParentHash,
+			"txs", len(engineEnvelope.ExecutionPayload.Transactions))
 	} else {
-		envelope, err = eng.GetPayload(ctx, payloadInfo)
+		engineEnvelope, builderEnvelope, err = getPayloadWithBuilderPayload(ctx, log, eng, payloadInfo, l2head, builderClient, metrics)
 	}
 	if err != nil {
 		// even if it is an input-error (unknown payload ID), it is temporary, since we will re-attempt the full payload building, not just the retrieval of the payload.
-		return nil, BlockInsertTemporaryErr, fmt.Errorf("failed to get execution payload: %w", err)
+		return nil, BlockInsertTemporaryErr, fmt.Errorf("failed to get execution payload from engine: %w", err)
+	}
+
+	if builderEnvelope != nil {
```

Comment:
more of a spec question: if builder profit < engine payload profit, should the engine payload be chosen over the builder payload? There needs to be a way to expose the engine block value instead of blindly choosing the builder envelope

Reply:
good point. It's not clear how op stack actually "profits" from sequencing. the user tx fees is a function of L1 gas fees and i am not sure if they take any margins from the difference. I just asked the optimism discord if there is a way to calculate the profits